### PR TITLE
Stepper: Track step-route unconditionally - logged-out/empty

### LINKED ISSUE #94715: Stepper: Track the initial step route access that redirects to login/User step
Currently, accessing a route such as `/setup/onboarding/plans` will skip logging/tracking if the step requires authentication and a redirection happens to `/setup/onboarding/user`. This is potentially a data point that is missing. 

The reasoning is that the initial request is part of the user experience/journey and we should probably Track/log it like normal:

From https://github.com/Automattic/wp-calypso/pull/94528#discussion_r1765348363:

> Agree for the most part, but also think that it's through the system logs that you can navigate the full extend of their experience. Stepper being user-facing framework, pretty much whatever it does is typically a reaction.
> 
> The initial request is also part of the user experience, no? So to understand their journey I'd be looking at the full spectrum below:
> 
> 1. User goes to `/setup/onboarding/plans`
> 2. Redirected to `/setup/onboarding/user`
> 3. Signs up
> 4. Rdirected to `/setup/onboarding/domains` (not plans)
> 
> Ultimately not accounting anything around 1 may be a data point missing. Insignificant as it may be/seem.
>

### PR DESCRIPTION
**Pull request by chriskmnds**
Fixes https://github.com/Automattic/wp-calypso/issues/94715

## Proposed Changes

We basically now always track a step route access irrespective of any conditions that may or may not render the step content. The conditions were so far:
- a step requires logging in. The user is redirected to `/setup/:flow/user` in that case
- a step does not render any content and a redirection happens
  - this usually results from either the step not having loaded yet or having `AssertConditionState.FAILURE === true` - the latter happens when requirements aren't met / such as requiring login, or missing parameters)

This PR essentially reverts a lot of the underlying logic in https://github.com/Automattic/wp-calypso/pull/91241. The motivation for doing so is in https://github.com/Automattic/wp-calypso/issues/94715 (i.e. just accessing a given step is a data point worth tracking). There was a bit of discussion in https://github.com/Automattic/wp-calypso/pull/91241 (cc @daledupreez) - so let's doubly confirm we are onboard with this direction :-)

## Why are these changes being made?

Addresses https://github.com/Automattic/wp-calypso/issues/94715

## Testing Instructions

* Go to `/setup/onboarding/plans` logged-out
* Confirm `calypso_signup_step_start`, `calypso_signup_actions_complete_step`, and `calypso_page_view` get recorded before navigated to `/setup/onboarding/user` for logging in
* Log in and progress through to `/setup/onboarding/plans` again. Confirm the same events being tracked

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Review by renancarvalho — COMMENTED** [review]
Thank you for working on it @chriskmnds 

I took a screenshot divided into two parts.

The events are being triggered, I am just not sure if the order is correct. Maybe it could be Tracks vigilante ordering them differently? Not sure.

Part 1
image

Part 2
image

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx [hunk 1]
```diff
@@ -21,34 +20,33 @@ import {
...
-// Ensure that the selected site is fetched, if available. This is used for event tracking purposes.
-// See https://github.com/Automattic/wp-calypso/pull/82981.
-const useIsRequestingSelectedSite = ( siteSlugOrId: string | number, site: SiteDetails | null ) => {
-	return useSelector( ( state ) => site && isRequestingSite( state, siteSlugOrId ) );
+/**
+ * We wait for the site to be fetched before tracking the step route when a site ID/slug are defined in the params.
+ * This is to ensure that the site data is available for event tracking purposes.
+ * See `superProps`, `site_plan_id`, and https://github.com/Automattic/wp-calypso/pull/82981.
+ */
+const useHasRequestedSelectedSite = () => {
```
Comment (renancarvalho) [inline client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx:28]: This function works well, I just had a bit of a hard time reading the return of it.

What do you think if we make it simpler?

Example 

```js
const useHasRequestedSelectedSite = () => {
	const { site, siteSlugOrId } = useSiteData();

	if (!siteSlugOrId) {
		return true;
	}

	const isRequesting = useSelector((state) => isRequestingSite(state, siteSlugOrId));

	return !!site && !isRequesting;
};
```

I haven't test, but I believe it is doing the same and IMO more readable. This is totally optional.

Comment (chriskmnds) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx:28]: Thanks. It felt same way to me too. We can't do an early return though before `useSelector` call (can't call hooks conditionally). 

Let me know if anything comes to mind. I think before it was a bit worse with being scattered around (but we could bring that version back too) 🤔

**Comment by chriskmnds** [comment]
> Thank you for working on it @chriskmnds
> 
> I took a screenshot divided into two parts.
> 
> The events are being triggered, I am just not sure if the order is correct. Maybe it could be Tracks vigilante ordering them differently? Not sure.
> 
> Part 1 image
> 
> Part 2 image

Thanks @renancarvalho ! It does look weird. These are all fired in the correct order. I think the timestamps may just not be "granular" enough. The redirection happens in the order of milliseconds and none of the dates in the snapshots differ. 

In Network tab, I get the correct order with same Date in headers:

- plans -> Mon, 23 Sep 2024 12:56:41 GMT
- user -> Mon, 23 Sep 2024 12:56:41 GMT

Definitely needs investigating. Not sure how they filter/order the events in funnels or if we can introduce a more granular timestamp on each event. Perhaps @travisw can clarify?

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx (hunk 1 above)
Comment (renancarvalho) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx:28]: True it is inside a hook. 
I've asked "help" to our good friend gpt.
It suggested.

```js
const useHasRequestedSelectedSite = () => {
  const { site, siteSlugOrId } = useSiteData();

  // Always call the hook but adjust logic within the selector
  const isRequestingSelectedSite = useSelector((state) => {
    if (!site || !siteSlugOrId) {
      return false; // No request if site or siteSlugOrId is missing
    }
    return isRequestingSite(state, siteSlugOrId);
  });

  // Memoize the result to prevent unnecessary re-renders
  return useMemo(() => {
    if (!siteSlugOrId) {
      return true; // If no siteSlugOrId, assume requested
    }
    return !!site && !isRequestingSelectedSite;
  }, [site, siteSlugOrId, isRequestingSelectedSite]);
};

```
I've tested and works the same, plus I liked the useMemo here.
Again, this is totally optional.

**Comment by chriskmnds** [comment]
> Definitely needs investigating. Not sure how they filter/order the events in funnels or if we can introduce a more granular timestamp on each event. Perhaps @travisw can clarify? 

I can see there is a `_ts: data.getTime()` prop in [recordEvent](https://github.com/Automattic/wp-calypso/blob/trunk/client/server/lib/analytics/index.js), which sounds like the timestamp we need. It also creates the right order in the snapshots if we follow those. 

@renancarvalho let's wait for @travisw to confirm, and it might be the case we need to update Tracks vigilante to order with that value instead 🤔

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx (hunk 1 above)
Comment (chriskmnds) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx:28]: Interesting. Thanks for probing the AI for it! :D 

I feel memoizing anything here isn't effective. The value is true/false, so unless we save on processing (we don't), then it's probably unnecessary?

Comment (renancarvalho) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx:28]: Think it could be useful, preventing unnecessary re-renderings, needs better validation, though.

**Comment by MicBosi** [comment]
I can't see the images [here](https://github.com/Automattic/wp-calypso/pull/94817#issuecomment-2368168212) but yes `_ts` is the most granular timestamp to be used. I'd use the Network tab as source of truth to see who is firing in what order if Tracks Vigilante's data looks dubious.

**Comment by MicBosi** [comment]
Having had a look at the "legacy" `/start`, it seems that it too fires `calypso_signup_step_start` and even `calypso_page_view` in cases where users _do not_ see the step due to redirection to login.

If Stepper avoids doing that, and rather provides mechanisms to distinguish when a precondition fails/redirect happens vs when a step is actually experienced by the user without introducing false positives, I think these would arguably all be extra points to Stepper in the area or data reliability over `/start`.

**Comment by chriskmnds** [comment]
> If upon initialisation of a step some preconditions fail perhaps we could have a dedicated Tracks event (eg calypso_signup_step_condition or similar), with some standard prop to distinguish the type of condition, ex. user not logged in / redirection to step X. This way we can also cleanly analyse how many times users get redirected between steps or which steps have which preconditions fail more often and potentially derive interesting insights from it.

@MicBosi would a property on `calypso_signup_step_start` work equally (if not better?) for this? I think redirections can happen more sporadically outside of these AssertConditions (such as the very case of redirecting on requiring login right now). Technically, nothing effectively guarantees that a user will see something upon navigating to a step. 

So do we need more Tracks events and conditioning the ones we have? I would think a couple more props would serve things better, like `step_content`, `assert_condition`, etc. I mean, isn't analysis done through these things?

**Comment by MicBosi** [comment]
Thanks @chriskmnds. I agree, if it's possible to use a prop to distinguish a `calypso_signup_step_start` that renders anything to the user vs ones that is either skipped or redirected or what not, then that works too and would still be an improvement over `/start`.

I'm not sure if by the time `calypso_signup_step_start` fires you'll have all the required information at hand to fill those props, but it doesn't have to be a perfect solution, if we can distinguish at least a few of the most common cases where a step it's obviously being skipped or redirected, then that's still a step towards more accurate data - feel free to only consider low hanging fruits, if there are any.

To be honest, I don't even feel this is a super critical issue that should block the start of the experiment anyway because, 1) worst case scenario we're back to what `/start` was doing, ie, not worse, 2) we can always improve later, 3) it does feel like a savvy approach would need a deeper conversation, some testing/analysis and more time than we have at the moment.

**Comment by chriskmnds** [comment]
Thanks @MicBosi 

> I'm not sure if by the time calypso_signup_step_start fires you'll have all the required information at hand to fill those props, 

We probably do have most of what matters right now. The location where the event gets recorded is where we decide whether to render or not (and there's not much dynamism going on there, right now).

> but it doesn't have to be a perfect solution, if we can distinguish at least a few of the most common cases where a step it's obviously being skipped or redirected, then that's still a step towards more accurate data - feel free to only consider low hanging fruits, if there are any.
> 
> To be honest, I don't even feel this is a super critical issue that should block the start of the experiment anyway because, 1) worst case scenario we're back to what /start was doing, ie, not worse, 2) we can always improve later, 3) it does feel like a savvy approach would need a deeper conversation, some testing/analysis and more time than we have at the moment.

Absolutely. A holistic solution to this, clarifying the reasons for a step not rendering will require greater effort. We don't have anything of the sort right now, outside of some kinda half-baked boolean assertion logic.

I was not aware that `/start` was already doing this. My main motivation is actually to not deviate too far off Start without good reason, so works both ways for me.

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx (hunk 1 above)
Comment (chriskmnds) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx:28]: I think memo here seems useless, especially as far as rerenders go.

The way I see it, memoizing has two main underlying purposes:

1. Avoid redoing expensive calculations
2. Ensure referential integrity on return values to avoid rerenders (this only concerns `Object` return values)

The return value here is a boolean `true`/`false`. There's no reference being recreated if this hook is called a thousand times and returns a thousand `true`. They are the same static value. Much like a string or a number or `null`/`undefined`.

**Comment by MicBosi** [comment]
> The location where the event gets recorded is where we decide whether to render or not

Nice, that's actually quite convenient.

> I was not aware that /start was already doing this.

To be honest, I had a hunch about it, but I wasn't 100% sure, which is why I tested it. For future reference and in case anybody wants to double check what I did to test is: 
- Open in incognito window https://wordpress.com/start/domains?ref=pricing-lp (which is where you get redirected to when clicking on "start" from the plans page if you were logged in).
- Calypso redirected me to https://wordpress.com/start/user-social as expected
- Check the Tracks events in the Network tab. To simplify, what I see is:
   - `calypso_signup_start` with 
      - `_dl: https://wordpress.com/start/domains?ref=pricing-lp` (document location)
   - `calypso_signup_step_start` with 
      - `step: domains`
      - `_dl: https://wordpress.com/start/domains?ref=pricing-lp`
   - `calypso_page_view` with 
      - `path: /start/domains`
      - `_dl: https://wordpress.com/start/domains?ref=pricing-lp`
   - `calypso_signup_step_start` with 
      - `step: user`
      - `_dl: https://wordpress.com/start/user-social`
   - `calypso_page_view` with 
      - `path: /start/user-social`
      - `_dl: https://wordpress.com/start/user-social`

However, this is just one case, I wouldn't be too surprised if in other redirection cases the picture looked different.

**Comment by chriskmnds** [comment]
Thanks @MicBosi + for the testing details!

I went ahead and introduced `skip_step_render` to all of `calypso_signup_step_start`, `calypso_page_view`, and `calypso_signup_actions_complete_step` calls, which are all tracking the step route.
- All of these should now record `true` when it's the redirect (e.g. visiting `/setup/onboarding/plans` logged out) and `false` otherwise. We can also update to _not_ send the prop when it's `false` (if for whatever reason that makes a difference)
- For the name, `skip_step_render`, it's more of a placeholder. Let me know if something else will be better and we can change.

p.s. Once we are happy with this, will we need to update/recreate the Tracks events to include the new prop? I'm not sure if that's necessary or not anymore https://mc.a8c.com/tracks/events/

**Comment by MicBosi** [comment]
> I went ahead and introduced skip_step_render to all of calypso_signup_step_start, calypso_page_view, and calypso_signup_actions_complete_step calls, which are all tracking the step route.

Thanks @chriskmnds! Looks great to me. Also the `skip_step_render` name sounds good accurate.

> We can also update to not send the prop when it's false (if for whatever reason that makes a difference)

TL;DR; yes, let's do that, ie not setting the prop if it'd be set to `false`.

That is a good question. I was going to say that it's nice to have it set to `false` when we know it's not rendering BUT on second thought, I think it's a bit safer on our data side if you simply don't set it, this way when we analyse the data we'll all use one way to check whether a step is being skipped, ie, the SQL equivalent of `skip_step_render !== true` in JS parlance - which is a check that is also backwards compatible with events fired by the "legacy" /start framework.

I want to avoid anyone checking for `skip_step_render === false` in SQL queries which would filter out all the non Stepper events.

> p.s. Once we are happy with this, will we need to update/recreate the Tracks events to include the new prop? I'm not sure if that's necessary or not anymore

I don't think you need to do anything. Do you have a specific example of what was done in the past?

**Comment by chriskmnds** [comment]
Thanks @MicBosi!

> TL;DR; yes, let's do that, ie not setting the prop if it'd be set to false.
> 
> That is a good question. I was going to say that it's nice to have it set to false when we know it's not rendering BUT on second thought, I think it's a bit safer on our data side if you simply don't set it, this way when we analyse the data we'll all use one way to check whether a step is being skipped, ie, the SQL equivalent of skip_step_render !== true in JS parlance - which is a check that is also backwards compatible with events fired by the "legacy" /start framework.
>
> I want to avoid anyone checking for skip_step_render === false in SQL queries which would filter out all the non Stepper events.

Yes, good thinking! It makes sense. Updated to skip sending it unless `true`.

> I don't think you need to do anything. Do you have a specific example of what was done in the past?

I vaguely recall several years back creating a PR to update an event with new metadata/props (the interface won't allow you to). I guess `skip_step_render` will show on the event page once recorded? We'll find out once we ship here :-) 

I'll run a few more tests and merge this a little later. Let me know if you get a chance to review again. We can also follow up on anything post-merge.

**Comment by chriskmnds** [comment]
I had to battle with an issue with `calypso_page_view` recording the previous `path` when redirecting from User to Domains after logging in. Apparently, this event is recorded behind a timeout and it messes things when the `window.location.pathname` call comes from inside the `useEffect` hook, as I found, moving this outside fixes the issue.

I've done several tests and it looks good to me to proceed. Hopefully I haven't missed something, but the values (especially `path` on `calypso_page_view`) look correct. So FYI @escapemanuele / @MicBosi / @renancarvalho if anything comes up
